python/sglang/srt/mem_cache/cache_controller_direct.py

```diff
@@ -0,0 +1,328 @@
+        device_id: int = 0,
+    ):
+        self.mem_pool_device_allocator = token_to_kv_pool_allocator
+        self.mem_pool_device = token_to_kv_pool_allocator.get_kvcache()
+        if self.mem_pool_device_allocator:
+            self.device = self.mem_pool_device_allocator.device
+        else:
+            self.device = torch.device("cpu")
+        # self.kv_layer_ptrs: every layer ptr
+        # self.kv_layer_nbytes: the byte length of each layer
+        # self.kv_page_nbytes: the page byte length of each layer
+        self.kv_layer_ptrs, self.kv_layer_nbytes, self.kv_page_nbytes = (
+            self.mem_pool_device.get_contiguous_buf_infos()
+        )
+
+        self.page_size = page_size
+        self.device_id = device_id
+        self.is_mla_model = isinstance(self.mem_pool_device, MLATokenToKVPool)
+
+        if is_dp_attention_enabled():
+            self.tp_rank = get_attention_tp_rank()
+            self.tp_size = get_attention_tp_size()
+            self.dp_rank = get_attention_dp_rank()
+        else:
+            self.tp_rank = get_tensor_model_parallel_rank()
+            self.tp_size = get_tensor_model_parallel_world_size()
+            self.dp_rank = 0
+
+        # for MLA models, only one rank needs to backup the KV cache
+        self.backup_skip = self.is_mla_model and self.tp_rank != 0
+
+        self.storage_config = HiCacheStorageConfig(
+            tp_rank=self.tp_rank,
+            tp_size=self.tp_size,
+            is_mla_model=self.is_mla_model,
+            is_page_first_layout=False,
+            model_name=None,
+            extra_config={"device_id": device_id},
+        )
+        try:
+            self.storage_backend = StorageBackendFactory.create_backend(
+                storage_backend, self.storage_config, None
+            )
+        except ValueError as e:
+            raise ValueError(f"Failed to create storage backend: {e}") from e
+
+        self.storage_backend.register_mem_pool_device(self.mem_pool_device)
+
+        self.load_tokens_threshold = 128
+        # granularity of batch storage IO operations, in number of pages
+        self.storage_batch_size = 256
+
+        # create a new communication group for synchronizing storage operations across TP workers
+        self.tp_world_size = torch.distributed.get_world_size(group=tp_group)
+        if self.tp_world_size > 1:
+            group_ranks = torch.distributed.get_process_group_ranks(tp_group)
+            self.load_tp_group = torch.distributed.new_group(
+                group_ranks, backend="gloo"
+            )
+
+        self.layer_num = self.mem_pool_device.layer_num
+        self.layer_done_counter = LayerDoneCounter(self.layer_num)
+        self.mem_pool_device.register_layer_transfer_counter(self.layer_done_counter)
+
+        self.load_queue: List[LoadStorageOperation] = []
+
+    def reset(self):
+        self.load_queue.clear()
+
+    def write(self, hash_keys: List[str], device_indices: torch.Tensor) -> int:
```

Comment:
in the old controller, operations were performed in separate thread and there was only queuing. It seems that io is blocking the scheduler thread right now, but maybe I'm wrong